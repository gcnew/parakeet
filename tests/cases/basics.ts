
import { test } from '../test_util'

import {
    char, string, number, ws,

    withPosition, getLineCol
} from '../../src/string_combinators'

import {
    map, many, combine, choice, separatedTrailing, separatedZeroTrailing,

    left, right, pair
} from '../../src/parser_combinators'


const ll = choice(
    [ string('num'), map(number, n => Number(n)) ],
    [ string('str'), string('hello') ]
);

const nn = map(number, x => +x);

const wp0 = combine(many(ws), withPosition(number), (_, x) => x);
const wp = (st: any) => {
    const lineOffsets = st.getData().lineOffsetTable;
    const result = wp0(st);
    const value = result.kind === 'left' ? result.value
                                         : result.value[0];


    const [line, col] = getLineCol(value.position, lineOffsets)!.map(x => x + 1);
    const retval = {
        line, col, value: value.value
    };

    if (result.kind === 'left') {
        return left(retval)
    }

    return right(pair(retval, result.value[1]));
}

const numList  = separatedTrailing(number, char(','));
const numList2 = separatedZeroTrailing(number, char(','));

test(ll, 'strhello');
test(ll, 'strhello world');
test(nn, '123abc');
test(wp, '\n\n123');
test(wp, '\n\n   hello?');

test(numList, '');
test(numList, '1');
test(numList, '1,');
test(numList, '1,2,3');
test(numList, '1,2,3,');

test(numList2, '');
test(numList2, '1');
test(numList2, '1,');
test(numList2, '1,2,3');
test(numList2, '1,2,3,');
