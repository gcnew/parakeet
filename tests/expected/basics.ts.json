{
    "kind": "right",
    "value": [
        "hello",
        {
            "source": "strhello",
            "offset": 8,
            "data": {
                "lineOffsetTable": [
                    [
                        0,
                        "strhello"
                    ]
                ]
            }
        }
    ]
}

{
    "kind": "right",
    "value": [
        "hello",
        {
            "source": "strhello world",
            "offset": 8,
            "data": {
                "lineOffsetTable": [
                    [
                        0,
                        "strhello world"
                    ]
                ]
            }
        }
    ]
}

{
    "kind": "right",
    "value": [
        123,
        {
            "source": "123abc",
            "offset": 3,
            "data": {
                "lineOffsetTable": [
                    [
                        0,
                        "123abc"
                    ]
                ]
            }
        }
    ]
}

{
    "kind": "right",
    "value": [
        {
            "line": 3,
            "col": 1,
            "value": "123"
        },
        {
            "source": "\n\n123",
            "offset": 5,
            "data": {
                "lineOffsetTable": [
                    [
                        0,
                        ""
                    ],
                    [
                        1,
                        ""
                    ],
                    [
                        2,
                        "123"
                    ]
                ]
            }
        }
    ]
}

{
    "kind": "left",
    "value": {
        "line": 3,
        "col": 4,
        "value": {
            "kind": "pc_error",
            "code": "digit_expected"
        }
    }
}

{
    "kind": "left",
    "value": {
        "kind": "pc_error",
        "code": "eos_reached"
    }
}

{
    "kind": "right",
    "value": [
        [
            "1"
        ],
        {
            "source": "1",
            "offset": 1,
            "data": {
                "lineOffsetTable": [
                    [
                        0,
                        "1"
                    ]
                ]
            }
        }
    ]
}

{
    "kind": "right",
    "value": [
        [
            "1"
        ],
        {
            "source": "1,",
            "offset": 2,
            "data": {
                "lineOffsetTable": [
                    [
                        0,
                        "1,"
                    ]
                ]
            }
        }
    ]
}

{
    "kind": "right",
    "value": [
        [
            "1",
            "2",
            "3"
        ],
        {
            "source": "1,2,3",
            "offset": 5,
            "data": {
                "lineOffsetTable": [
                    [
                        0,
                        "1,2,3"
                    ]
                ]
            }
        }
    ]
}

{
    "kind": "right",
    "value": [
        [
            "1",
            "2",
            "3"
        ],
        {
            "source": "1,2,3,",
            "offset": 6,
            "data": {
                "lineOffsetTable": [
                    [
                        0,
                        "1,2,3,"
                    ]
                ]
            }
        }
    ]
}

{
    "kind": "right",
    "value": [
        [],
        {
            "source": "",
            "offset": 0,
            "data": {
                "lineOffsetTable": [
                    [
                        0,
                        ""
                    ]
                ]
            }
        }
    ]
}

{
    "kind": "right",
    "value": [
        [
            "1"
        ],
        {
            "source": "1",
            "offset": 1,
            "data": {
                "lineOffsetTable": [
                    [
                        0,
                        "1"
                    ]
                ]
            }
        }
    ]
}

{
    "kind": "right",
    "value": [
        [
            "1"
        ],
        {
            "source": "1,",
            "offset": 2,
            "data": {
                "lineOffsetTable": [
                    [
                        0,
                        "1,"
                    ]
                ]
            }
        }
    ]
}

{
    "kind": "right",
    "value": [
        [
            "1",
            "2",
            "3"
        ],
        {
            "source": "1,2,3",
            "offset": 5,
            "data": {
                "lineOffsetTable": [
                    [
                        0,
                        "1,2,3"
                    ]
                ]
            }
        }
    ]
}

{
    "kind": "right",
    "value": [
        [
            "1",
            "2",
            "3"
        ],
        {
            "source": "1,2,3,",
            "offset": 6,
            "data": {
                "lineOffsetTable": [
                    [
                        0,
                        "1,2,3,"
                    ]
                ]
            }
        }
    ]
}
